### PR TITLE
vmware_guest: split in 3 modules (guest, guest_facts & guest_snapshot)

### PR DESCRIPTION
**Pull request by nerzhul**
##### ISSUE TYPE
 - Feature Pull Request
 - New Module Pull Request

##### COMPONENT NAME
vmware_guest
vmware_guest_vm_facts
vmware_guest_snapshot

##### ANSIBLE VERSION
```
ansible 2.3 devel
```

##### SUMMARY

Split vmware_guest in 3 parts: vmware_guest, vmware_guest_facts, vmware_guest_snapshot

New module: vmware_guest_facts

* Move gather_facts action outside of vmware_guest module
* Move facts gathering method to vmware.py library
* vmware_guest use some vmware.py library methods
* vmware.py lib: fix some PEP 8 issues
* facts now returns snapshot list if available

New module: vmware_guest_snapshot

* Create, remove(all), revert snapshots

### CODE REVIEW

**Comment by nerzhul** [comment]
@dagwieers i moved facts outside of vmware_guest module while hacking at ansible meetup in paris

cc @gundalow

**Comment by dagwieers** [comment]
You rule ! Let's meet !

**Comment by dagwieers** [comment]
Why vmware_guest_vm_facts and not vmware_guest_facts.
Are there any non-vm guests in vnware ?

**Comment by nerzhul** [comment]
@dagwieers you are right, it's too long, changing it now and fixing changelog.md and looking for shippable if it wants to load on this wifi :(

**Comment by nerzhul** [comment]
@dagwieers rebased, force pushed, renamed :)
seems @ansibot is at Ansible Automates event and doesn't want to tag :(

**Comment by nerzhul** [comment]
seems @ansibot is at Ansible Automates event and doesn't want to tag :(

**Review by dagwieers — APPROVED** [review]
Cannot test, but since this is mostly relocating existing code, I am confident this is quite safe.

**Comment by nerzhul** [comment]
@dagwieers i added a second commit and moved snapshot part to a third module to manage snapshots only, without the listing parts, which needs more tuning

**Comment by nerzhul** [comment]
It's now splited in four parts and need testing @dagwieers when you are ready. I will test all tomorrow more heavily at work

**Comment by dagwieers** [comment]
Is there really a need to have 4 different modules ?
Can't we leave the snapshot facts as part of the vmware_guest_facts ?

**Comment by nerzhul** [comment]
@dagwieers i can merge the snapshot facts into the VM itself, it's possible yes, if needed. @dav1x @jctanner what do you think about this case ?

**Comment by dagwieers** [comment]
Given the fact that so many *_facts modules already exist for other large and small subsystems I don't mind if this get merged as-is. However I hope that all these *_facts modules eventually and up being facts plugins in a new pluggable facts mechanism that we still have to design :-)

**Comment by jctanner** [comment]
@nerzhul yeah, let's combine the two facts modules to reduce confusion for the end users.

**Review by jctanner — DISMISSED** [review]
please combine the facts modules

File: lib/ansible/module_utils/vmware.py [hunk 1]
```diff
@@ -181,6 +185,69 @@ def find_host_portgroup_by_name(host, portgroup_name):
     return None
 
 
+def gather_vm_facts(content, vm):
```
Comment (ritzk) [inline lib/ansible/module_utils/vmware.py:188]: Is gather_vm_facts used by any other module ?

**Comment by nerzhul** [comment]
Yes, it's used by VMware guest when doing actions on a machine and the guest facts

**Comment by KroArtem** [comment]
While you're at it, can you please fix this one: https://github.com/ansible/ansible/issues/20070
It's just a ctrl-c, ctrl-v action. :)

**Comment by nerzhul** [comment]
WIll do the module merge and re-check the snapshot problem today at work if i'm not too busy

**Comment by KroArtem** [comment]
Thanks!

**Comment by nerzhul** [comment]
@gundalow @dagwieers @KroArtem @jctanner all concerns are now fixes.

We now have 3 modules. I re-tested and do some fixed on vmware_guest_snapshot, it works perfect on my 5.5 cluster. I tested, add 10 snapshot, revert, re-revert, remove a snapshot in the middle, the first snapshot, remove_all, everything is okay (except my machine which became crazy after many snapshot changes operations :p but after restart it's okay

**Comment by dagwieers** [comment]
Looking good to me !
shipit

File: lib/ansible/module_utils/vmware.py (hunk 1 above)
Comment (dagwieers) [reply lib/ansible/module_utils/vmware.py:188]: Yes it is. (Answering the review)

**Comment by jctanner** [comment]
@nerzhul looks like you have a few code style problems ...

```
2017-02-06 10:13:57 Run command: pep8 --max-line-length 160 --config /dev/null --ignore E123,E124,E127,E128,E201,E202,E203,E211,E221,E222,E225,E226,E227,E228,E231,E241,E251,E261,E262,E265,E266,E301,E302,E303,E402,E502,E713,E731,W391,W503 lib/ansible/module_utils/vmware.py lib/ansible/modules/cloud/vmware/vmware_guest.py lib/ansible/modules/cloud/vmware/vmware_guest_facts.py lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py
2017-02-06 10:13:58 ERROR: PEP 8: lib/ansible/module_utils/vmware.py: Passes current rule set. Remove from legacy list (test/sanity/pep8/legacy-files.txt).
2017-02-06 10:13:58 PEP 8: There were 1 different legacy issues found (3 total):
2017-02-06 10:13:58 PEP 8: Count Code Message
2017-02-06 10:13:58 PEP 8:     3 E501 line too long (NNN > NNN characters)
2017-02-06 10:13:58 ERROR: PEP 8: There are 1 issues which need to be resolved.
2017-02-06 10:13:58 ERROR: InvocationError: '/root/src/github.com/ansible/ansible/test/runner/test.py sanity -v --skip-test ansible-doc --python 2.7 --require CHANGELOG.md --require lib/ansible/module_utils/vmware.py --require lib/ansible/modules/cloud/vmware/vmware_guest.py --require lib/ansible/modules/cloud/vmware/vmware_guest_facts.py --require lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py --color yes --requirements'
2017-02-06 10:13:58 ___________________________________ summary ____________________________________
2017-02-06 10:13:58 ERROR:   py27: commands failed
2017-02-06 10:13:58 ERROR: Command "tox -c test/runner/tox.ini -e py27 -- /root/src/github.com/ansible/ansible/test/runner/test.py sanity -v --skip-test ansible-doc --python 2.7 --require CHANGELOG.md --require lib/ansible/module_utils/vmware.py --require lib/ansible/modules/cloud/vmware/vmware_guest.py --require lib/ansible/modules/cloud/vmware/vmware_guest_facts.py --require lib/ansible/modules/cloud/vmware/vmware_guest_snapshot.py --color yes --requirements" returned exit status 1.
2017-02-06 10:13:58 + cleanup
2017-02-06 10:13:58 ++ ls test/results/coverage/
2017-02-06 10:13:58 + '[' '' ']'
2017-02-06 10:13:58 + rmdir shippable/testresults/
2017-02-06 10:13:58 + cp -av test/results/junit/ shippable/testresults/
2017-02-06 10:13:58 ‘test/results/junit/’ -> ‘shippable/testresults/’
2017-02-06 10:13:58 ‘test/results/junit/.keep’ -> ‘shippable/testresults/.keep’
```

**Comment by gundalow** [comment]
2017-02-06 10:13:58 ERROR: PEP 8: lib/ansible/module_utils/vmware.py: Passes current rule set. Remove from legacy list (test/sanity/pep8/legacy-files.txt).

Is the issue. Please delete `lib/ansible/module_utils/vmware.pl` from `test/sanity/pep8/legacy-files.txt`

**Pull request merged**